src/gt4py/next/iterator/type_inference.py

```diff
@@ -380,20 +385,54 @@ def _infer_shift_location_types(shift_args, offset_provider, constraints):
     return current_loc_in, current_loc_out
 
 
-class _TypeInferrer(eve.NodeTranslator):
+@dataclasses.dataclass
+class _TypeInferrer(eve.traits.VisitorWithSymbolTableTrait, eve.NodeTranslator):
     """Visit the full iterator IR tree, convert nodes to respective types and generate constraints."""
 
-    def __init__(self, offset_provider):
-        self.offset_provider = offset_provider
+    offset_provider: Optional[dict[str, Connectivity | Dimension]]
+    collected_types: dict[int, Type] = dataclasses.field(default_factory=dict)
+    constraints: set[tuple[Type, Type]] = dataclasses.field(default_factory=set)
 
-    def visit_SymRef(
-        self, node: ir.SymRef, constraints: set[tuple[Type, Type]], symtypes: dict[str, Type]
-    ) -> Type:
-        if node.id in symtypes:
-            res = symtypes[node.id]
-            if isinstance(res, LetPolymorphic):
-                return freshen(res.dtype)
-            return res
+    @classmethod
+    def apply(
+        cls,
+        node: ir.Node,
+        offset_provider: Optional[dict[str, Connectivity | Dimension]] = None,
+        reindex: bool = True,
+    ) -> dict[int, Type]:
+        """
+        Infer the types of the child expressions of a given iterator IR expression.
+
+        The result is a dictionary mapping the id of child nodes to their type.
```

Comment:
Just a suggestion, it could be worth to make clear here that by `id` is meant the Python `id()` 
```suggestion
        The result is a dictionary mapping the Python id of child nodes to their type.
```